### ISSUE TITLE
ADIOS 2 IO (dev branch only)

### ISSUE THREAD

**Issue by mathrack** [2decomp-fft/2decomp-fft#338]
Since PR https://github.com/2decomp-fft/2decomp-fft/pull/335, CI is no longer running the ADIOS2 IO test.

I have attached below a minimal example allowing to reproduce the bug. If we can fix the example we can fix the CI or the code. Currently, it is not clear if the error is located in the CI example or in the IO part of 2decomp.

Running the program below on a single rank on my workstation using GNU compiler 11.2.0, OpenMPI 5.0.1 and ADIOS2 2.10.0, the output is : 
```
$ ./test 
 Registering variable u1.dat         275             16655200
 Registering variable u2.dat         276             16653712
 Writing variable out, u1.dat         292             16655200
 Writing variable out, u2.dat         293             16653712
 Reading variable out, u1.dat         295             16655200
 Error at line          153           4
STOP -1
```
The source of the program : 
```
! mpif90 -cpp -I/home/xxxx/adios2/build_2.10.0_openmpi_5.0.1_debug/include/adios2/fortran -g -Og test.f90 -Wl,-rpath,/home/xxxx/adios2/build_2.10.0_openmpi_5.0.1_debug/lib /home/xxxx/adios2/build_2.10.0_openmpi_5.0.1_debug/lib/libadios2_fortran_mpi.so.2.10.0 /home/xxxx/adios2/build_2.10.0_openmpi_5.0.1_debug/lib/libadios2_fortran.so.2.10.0 -Wl,-rpath-link,/home/xxxx/adios2/build_2.10.0_openmpi_5.0.1_debug/lib -o test

module tools

   use adios2

   implicit none

   private
   public :: check_error, &
             tools_register_var, &
             tools_write, &
             tools_read, &
             tools_print_adios2

   interface tools_print_adios2
      module procedure print_adios2_adios
      module procedure print_adios2_io
      module procedure print_adios2_engine
      module procedure print_adios2_variable
      module procedure print_adios2_namestruct
   end interface tools_print_adios2

contains

   ! Check the providded error code
   subroutine check_error(ierr, line)

      implicit none

      integer, intent(in) :: ierr, line

      if (ierr == 0) return

      write (*, *) "Error at line ", line, ierr
      stop - 1

   end subroutine check_error

   ! Register the given variable
   subroutine tools_register_var(adio, varname, sizes, line)

      implicit none

      ! Arguments
      type(adios2_io), intent(inout) :: adio
      character(len=*), intent(in) :: varname
      integer, intent(in) :: sizes(3), line

      ! Parameters
      integer, parameter :: data_type = adios2_type_dp
      integer, parameter :: ndims = 3
      logical, parameter :: adios2_constant_dims = .true.

      type(adios2_variable) :: var_handle
      integer :: ierror

      write (*, *) "Registering variable ", varname, line, adio%f2c

      call adios2_inquire_variable(var_handle, adio, varname, ierror)
      if (.not. var_handle%valid) then
         call adios2_define_variable(var_handle, &
                                     adio, &
                                     varname, &
                                     data_type, &
                                     ndims, &
                                     int(sizes, kind=8), &
                                     int((/0, 0, 0/), kind=8), &
                                     int(sizes, kind=8), &
                                     adios2_constant_dims, &
                                     ierror)
         call check_error(ierror, __LINE__)
      else
         write (*, *) "Error at line ", line, ierror
         stop - 2
      end if

   end subroutine tools_register_var

   ! Write the given array
   subroutine tools_write(adio, dirname, varname, array, line)

      implicit none

      ! Arguments
      type(adios2_io), intent(in) :: adio
      character(len=*), intent(in) :: dirname, varname
      double precision, dimension(:, :, :), intent(in) :: array
      integer, intent(in) :: line

      ! Parameters
      integer, parameter :: mode = adios2_mode_deferred
      integer, parameter :: access_mode = adios2_mode_write

      type(adios2_engine) :: engine
      type(adios2_variable) :: var_handle
      integer :: ierror

      write (*, *) "Writing variable ", dirname, ", ", varname, line, adio%f2c

      call adios2_open(engine, adio, dirname, access_mode, ierror)
      call check_error(ierror, __LINE__-1)

      call adios2_begin_step(engine, ierror)
      call check_error(ierror, __LINE__-1)

      call adios2_inquire_variable(var_handle, adio, varname, ierror)
      call check_error(ierror, __LINE__-1)

      call adios2_put(engine, var_handle, array, mode, ierror)
      call check_error(ierror, __LINE__-1)

      if (mode == adios2_mode_deferred) then
         call adios2_perform_puts(engine, ierror)
         call check_error(ierror, __LINE__-1)
      end if

      call adios2_end_step(engine, ierror)
      call check_error(ierror, __LINE__-1)

      call adios2_close(engine, ierror)
      call check_error(ierror, __LINE__-1)

   end subroutine tools_write

   ! Read the given array
   subroutine tools_read(adio, dirname, varname, array, line)

      implicit none

      ! Arguments
      type(adios2_io), intent(in) :: adio
      character(len=*), intent(in) :: dirname, varname
      double precision, dimension(:, :, :), intent(out) :: array
      integer, intent(in) :: line

      ! Parameters
      integer, parameter :: mode = adios2_mode_deferred
      integer, parameter :: access_mode = adios2_mode_read

      type(adios2_engine) :: engine
      type(adios2_variable) :: var_handle
      integer :: ierror

      write (*, *) "Reading variable ", dirname, ", ", varname, line, adio%f2c

      call adios2_open(engine, adio, dirname, access_mode, ierror)
      call check_error(ierror, __LINE__-1)

      call adios2_begin_step(engine, ierror)
      call check_error(ierror, __LINE__-1)

      call adios2_inquire_variable(var_handle, adio, varname, ierror)
      call check_error(ierror, __LINE__-1)

      call adios2_get(engine, var_handle, array, mode, ierror)
      call check_error(ierror, __LINE__-1)

      if (mode == adios2_mode_deferred) then
         call adios2_perform_gets(engine, ierror)
         call check_error(ierror, __LINE__-1)
      end if

      call adios2_end_step(engine, ierror)
      call check_error(ierror, __LINE__-1)

      call adios2_close(engine, ierror)
      call check_error(ierror, __LINE__-1)

   end subroutine tools_read

   subroutine print_adios2_adios(ad)

      implicit none

      type(adios2_adios), intent(in) :: ad

      write(*,*) "adios2_adios valid : ", ad%valid
      write(*,*) "             f2c   : ", ad%f2c

   end subroutine print_adios2_adios

   subroutine print_adios2_io(adio)

      implicit none

      type(adios2_io), intent(in) :: adio

      write(*,*) "adios2_io valid        : ", adio%valid
      write(*,*) "          engine_type  : ", adio%engine_type
      write(*,*) "          f2c          : ", adio%f2c

   end subroutine print_adios2_io

   subroutine print_adios2_engine(eng)

      implicit none

      type(adios2_engine), intent(in) :: eng

      write(*,*) "adios2_engine name  : ", trim(eng%name)
      write(*,*) "              type  : ", eng%type
      write(*,*) "              mode  : ", eng%mode
      write(*,*) "              valid : ", eng%valid
      write(*,*) "              f2c   : ", eng%f2c

   end subroutine print_adios2_engine

   subroutine print_adios2_variable(var)

      implicit none

      type(adios2_variable), intent(in) :: var

      write(*,*) "adios2_variable name  : ", trim(var%name)
      write(*,*) "                type  : ", var%type
      write(*,*) "                ndims : ", var%ndims
      write(*,*) "                valid : ", var%valid
      write(*,*) "                f2c   : ", var%f2c

   end subroutine print_adios2_variable

   subroutine print_adios2_namestruct(namestr)

      implicit none

      type(adios2_namestruct), intent(in) :: namestr

      write(*,*) "adios2_namestruct valid        : ", namestr%valid
      write(*,*) "                  count        : ", namestr%count
      write(*,*) "                  max_name_len : ", namestr%max_name_len
      write(*,*) "                  f2c          : ", namestr%f2c

   end subroutine print_adios2_namestruct

end module tools

program test

   use mpi
   use adios2
   use tools

   implicit none

   ! Parameters
   integer, parameter :: nx = 2, ny = nx + 1, nz = ny + 1
   character(len=*), parameter :: adios_xml = "adios2_config.xml"
   character(len=*), parameter :: dirout = "out"
   character(len=*), parameter :: file1 = "u1.dat"
   character(len=*), parameter :: file2 = "u2.dat"
   character(len=*), parameter :: adio1_label = "default"
   character(len=*), parameter :: adio2_label = "special"

   ! Adios2 objects
   type(adios2_adios), target :: adios
   type(adios2_io) :: adio1, adio2
   type(adios2_namestruct) :: namestr
   character(len=64) :: namelst(2)

   double precision, dimension(nx, ny, nz) :: arrayA, arrayB, arrayC, arrayD
   integer :: i, j, k, ierror

   ! Init MPI
   call MPI_init(ierror); call check_error(ierror, __LINE__)

   ! Init ADIOS2
   call adios2_init(adios, adios_xml, MPI_COMM_WORLD, ierror); call check_error(ierror, __LINE__)

   ! Init ADIOS2 IO families
   call adios2_declare_io(adio1, adios, adio1_label, ierror); call check_error(ierror, __LINE__)
   call adios2_declare_io(adio2, adios, adio2_label, ierror); call check_error(ierror, __LINE__)

   ! Register variables
   call tools_register_var(adio1, file1, (/nx, ny, nz/), __LINE__)
   call tools_register_var(adio2, file2, (/nx, ny, nz/), __LINE__)

   ! Init arrays
   do k = 1, nz
      do j = 1, ny
         do i = 1, nx
            arrayA(i, j, k) = i + (j - 1) * nx + (k - 1) * nx * ny
         end do
      end do
   end do
   arrayB = -arrayA

   if (.true.) then
      ! Write + write then read + read
      ! Error in the first read in adios2_inquire_variable

      call tools_write(adio1, dirout, file1, arrayA, __LINE__)
      call tools_write(adio2, dirout, file2, arrayB, __LINE__)

      call tools_read(adio1, dirout, file1, arrayC, __LINE__)
      call tools_read(adio2, dirout, file2, arrayD, __LINE__)

   else
      ! Write + read then write + read

      call tools_write(adio1, dirout, file1, arrayA, __LINE__)
      call tools_read(adio1, dirout, file1, arrayC, __LINE__)

      call tools_write(adio2, dirout, file2, arrayB, __LINE__)
      call tools_read(adio2, dirout, file2, arrayD, __LINE__)

   end if

   ! Error check
   if (maxval(abs(arrayA - arrayC)) > epsilon(arrayA(1, 1, 1)) .or. &
       maxval(abs(arrayB - arrayD)) > epsilon(arrayB(1, 1, 1))) then
      write (*, *) "Broken IO"
      stop 123456
   else
      write (*,*) "No error detected"
   end if

   ! Fin ADIOS2 IO families
   call adios2_flush_all_engines(adio1, ierror); call check_error(ierror, __LINE__)
   call adios2_remove_all_variables(adio1, ierror); call check_error(ierror, __LINE__)
   call adios2_remove_all_attributes(adio1, ierror); call check_error(ierror, __LINE__)
   call adios2_flush_all_engines(adio2, ierror); call check_error(ierror, __LINE__)
   call adios2_remove_all_variables(adio2, ierror); call check_error(ierror, __LINE__)
   call adios2_remove_all_attributes(adio2, ierror); call check_error(ierror, __LINE__)

   ! Fin ADIOS2
   call adios2_finalize(adios, ierror); call check_error(ierror, __LINE__)

   ! Fin MPI
   call MPI_finalize(ierror); call check_error(ierror, __LINE__)

end program test
```
The `adios2_config.xml` file : 
```
<?xml version="1.0"?>
<adios-config>
  <io name="special">
    <engine type="BP4">
    </engine>
    <transport type="File">
      <parameter key="Library" value="fstream"/>
      <parameter key="ProfileUnits" value="Milliseconds"/>
    </transport>
  </io>
  <io name="default">
    <engine type="BP4">
    </engine>
    <transport type="File">
      <parameter key="Library" value="fstream"/>
      <parameter key="ProfileUnits" value="Milliseconds"/>
    </transport>
  </io>
</adios-config>
```

**Comment by pbartholomew08** [comment]
The issue appears to be trying to write to the same location from different IO objects so the second call overwrites the first. I think the confusion comes from it being called `outdir` or similar - this was to try and have a unified implementation across MPIIO and ADIOS2, for all intents and purposes `outdir` is the output file in ADIOS2, and it replicates the behaviour in MPIIO:
```
# MPIIO
outdir/
- u.dat
- v.dat
- ...

# ADIOS2-BP4/5
outdir/
- database: u.dat, v.dat, ...

# ADIOS2-HDF5
outdir.h5
# Contains
  u.dat
  v.dat
  ...
```
  
I was able to run the code by changing the main to look like
```
   if (.true.) then
      ! Write + write then read + read
      ! Error in the first read in adios2_inquire_variable

      call tools_write(adio1, dirout//"1", file1, arrayA, __LINE__)
      call tools_write(adio2, dirout//"2", file2, arrayB, __LINE__)
      
      call tools_read(adio1, dirout//"1", file1, arrayC, __LINE__)
      call tools_read(adio2, dirout//"2", file2, arrayD, __LINE__)
                                     
   else                              
      ! Write + read then write + read
                                     
      call tools_write(adio1, dirout//"1", file1, arrayA, __LINE__)
      call tools_read(adio1, dirout//"2", file1, arrayC, __LINE__)
                                     
      call tools_write(adio2, dirout//"1", file2, arrayB, __LINE__)
      call tools_read(adio2, dirout//"2", file2, arrayD, __LINE__)
         
   end if
 ```
 
I'm not sure if this is what's intended.

You can check the file corruption by running the original code and
```
> $ADIOS2/bin/bpls out
  double   u2.dat  {4, 3, 2}
```
shows that the second call in the original code has overwritten the first (`u1.dat` is missing)

**Comment by pbartholomew08** [comment]
Also those string objects aren't available in older ADIOS2 (v2.7), I'm not sure when they get introduced so I had to comment these out

**Comment by mathrack** [comment]
Thank you for your help, it makes sense. I have started to work on fix based on this explanation
